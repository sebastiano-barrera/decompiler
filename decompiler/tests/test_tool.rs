//! Tests based on "test_tool"
//!
//! These are pretty weak, but they work on real code. They only check that the
//! decompiler doesn't panic, and they write the decompiler's output as text to
//! a version-controlled file, just so that it's somewhere easy to diff and hard
//! to forget updating.

use include_dir::{include_dir, Dir};
use insta::assert_snapshot;

use std::{
    path::Path,
    sync::{Mutex, MutexGuard, OnceLock},
};

static DATA_DIR: Dir<'_> = include_dir!("$CARGO_MANIFEST_DIR/test-data/");

macro_rules! case {
    ($case:ident, $exe:expr, $func_name:ident) => {
        #[test_log::test]
        #[allow(non_snake_case)]
        fn $case() {
            let function_name = stringify!($func_name);
            let out = $exe.process_function(function_name);
            crate::assert_snapshot!(out);
        }
    };
}

macro_rules! tests_in_binary {
    ($group:ident, $exe_path:expr; $($funcs:ident),* $(,)?) => {
        mod $group {
            use crate::Exe;
            static EXE: Exe = Exe::new($exe_path);

            $(case!($funcs, EXE, $funcs);)*
        }
    };
}

struct Exe {
    once_lock: OnceLock<Mutex<decompiler::Executable<'static>>>,
    path: &'static str,
}
impl Exe {
    const fn new(path: &'static str) -> Self {
        Exe {
            once_lock: OnceLock::new(),
            path,
        }
    }

    fn get_or_init(&self) -> MutexGuard<'_, decompiler::Executable<'static>> {
        let mutex = self.once_lock.get_or_init(|| {
            let rel_path = Path::new(self.path);
            let raw = DATA_DIR.get_file(rel_path).unwrap().contents();
            let exe = decompiler::Executable::parse(raw).unwrap();
            Mutex::new(exe)
        });
        mutex.lock().unwrap()
    }

    fn process_function(&self, function_name: &str) -> String {
        use std::fmt::Write;

        let mut exe = self.get_or_init();

        let df = exe
            .decompile_function(function_name)
            .expect("decompiling function");

        let mut log_buf = String::new();

        if let Some(mil) = df.mil() {
            writeln!(log_buf, " --- mil").unwrap();
            writeln!(log_buf, "{:?}\n", mil).unwrap();
        }

        if let Some(ssa) = df.ssa_pre_xform() {
            writeln!(log_buf, " --- ssa pre-xform").unwrap();
            writeln!(log_buf, "{:?}\n", ssa).unwrap();
        }

        if let Some(ssa) = df.ssa() {
            writeln!(log_buf, " --- cfg").unwrap();
            let cfg = ssa.cfg();
            writeln!(log_buf, "  entry: {:?}", cfg.direct().entry_bid()).unwrap();
            for bid in cfg.block_ids() {
                let regs: Vec<_> = ssa.block_regs(bid).collect();
                writeln!(
                    log_buf,
                    "  {:?} -> {:?} {:?}",
                    bid,
                    cfg.block_cont(bid),
                    regs
                )
                .unwrap();
            }
            write!(log_buf, "  domtree:\n    ").unwrap();

            let mut pp_buf = decompiler::pp::FmtAsIoUTF8(&mut log_buf);
            let pp = &mut decompiler::pp::PrettyPrinter::start(&mut pp_buf);
            cfg.dom_tree().dump(pp).unwrap();
            writeln!(log_buf).unwrap();

            writeln!(log_buf, " --- ssa").unwrap();
            writeln!(log_buf, "{:?}\n", ssa).unwrap();

            writeln!(log_buf, " --- ast").unwrap();
            let mut ast = decompiler::Ast::new(&ssa, exe.types());
            let mut pp_buf = decompiler::pp::FmtAsIoUTF8(&mut log_buf);
            let pp = &mut decompiler::pp::PrettyPrinter::start(&mut pp_buf);
            ast.pretty_print(pp).unwrap();
        }

        log_buf
    }
}

tests_in_binary!(
    exe_composite_type, "ty/test_composite_type.so";
    list_len
);
tests_in_binary!(exe_capstone, "integration/sample_capstone"; main);
tests_in_binary!(exe_matrix, "integration/sample_matrix"; sum_matrix);
tests_in_binary!(
    redis_server, "integration/redis-server";
    ctl_lookup,
    listNext,
    listTypePush,
    geoArrayCreate,
    // regression: infinite loop
    processAnnotations,
    redisReconnect,
    redisConnectWithOptions,
);
