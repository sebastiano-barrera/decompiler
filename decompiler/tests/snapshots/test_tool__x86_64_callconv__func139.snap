---
source: decompiler/tests/test_tool.rs
expression: out
---
func139 0x4013d2+11 (file 0x13d2)
          4013d2: 48 8b 44 24 18          mov rax,[rsp+18h]
          4013d7: 48 8b 54 24 20          mov rdx,[rsp+20h]
          4013dc: c3                      ret

function type: "func139" func (arg0 *void <TypeID(1v1)>,
                               arg1 *void <TypeID(1v1)>,
                               arg2 *void <TypeID(1v1)>,
                               arg3 *void <TypeID(1v1)>,
                               arg4 *void <TypeID(1v1)>,
                               arg5 small <TypeID(784v1)>,
                               arg6 small <TypeID(784v1)>) small <TypeID(784v1)>
0 warnings.


mil program = 
program  103 instrs
                 0 r12 <- Ancestral(AncestralName("stack_bottom"))
                 1 r2 <- Ancestral(AncestralName("CF"))
                 2 r3 <- Ancestral(AncestralName("PF"))
                 3 r4 <- Ancestral(AncestralName("AF"))
                 4 r5 <- Ancestral(AncestralName("ZF"))
                 5 r6 <- Ancestral(AncestralName("SF"))
                 6 r7 <- Ancestral(AncestralName("TF"))
                 7 r8 <- Ancestral(AncestralName("IF"))
                 8 r9 <- Ancestral(AncestralName("DF"))
                 9 r10 <- Ancestral(AncestralName("OF"))
                10 r11 <- Ancestral(AncestralName("RBP"))
                11 r12 <- Ancestral(AncestralName("RSP"))
                12 r13 <- Ancestral(AncestralName("RIP"))
                13 r14 <- Ancestral(AncestralName("RDI"))
                14 r15 <- Ancestral(AncestralName("RSI"))
                15 r16 <- Ancestral(AncestralName("RAX"))
                16 r17 <- Ancestral(AncestralName("RBX"))
                17 r18 <- Ancestral(AncestralName("RCX"))
                18 r19 <- Ancestral(AncestralName("RDX"))
                19 r20 <- Ancestral(AncestralName("R8"))
                20 r21 <- Ancestral(AncestralName("R9"))
                21 r22 <- Ancestral(AncestralName("R10"))
                22 r23 <- Ancestral(AncestralName("R11"))
                23 r24 <- Ancestral(AncestralName("R12"))
                24 r25 <- Ancestral(AncestralName("R13"))
                25 r26 <- Ancestral(AncestralName("R14"))
                26 r27 <- Ancestral(AncestralName("R15"))
                27 r28 <- Ancestral(AncestralName("ZMM0"))
                28 r29 <- Ancestral(AncestralName("ZMM1"))
                29 r30 <- Ancestral(AncestralName("ZMM2"))
                30 r31 <- Ancestral(AncestralName("ZMM3"))
                31 r32 <- Ancestral(AncestralName("ZMM4"))
                32 r33 <- Ancestral(AncestralName("ZMM5"))
                33 r34 <- Ancestral(AncestralName("ZMM6"))
                34 r35 <- Ancestral(AncestralName("ZMM7"))
                35 r36 <- Ancestral(AncestralName("ZMM8"))
                36 r37 <- Ancestral(AncestralName("ZMM9"))
                37 r38 <- Ancestral(AncestralName("ZMM10"))
                38 r39 <- Ancestral(AncestralName("ZMM11"))
                39 r40 <- Ancestral(AncestralName("ZMM12"))
                40 r41 <- Ancestral(AncestralName("ZMM13"))
                41 r42 <- Ancestral(AncestralName("ZMM14"))
                42 r43 <- Ancestral(AncestralName("ZMM15"))
                43 r44 <- Ancestral(AncestralName("memory"))
                44 r45 <- Undefined
                45 r46 <- Undefined
                46 r47 <- Undefined
                47 r48 <- Undefined
                48 r49 <- Undefined
                49 r50 <- Undefined
                50 r51 <- Undefined
                51 r52 <- Undefined
                52 r53 <- Undefined
                53 r54 <- Undefined
                54 r55 <- Undefined
                55 r56 <- Undefined
                56 r57 <- Undefined
                57 r58 <- Undefined
                58 r59 <- Undefined
                59 r60 <- Undefined
                60 r61 <- Undefined
                61 r62 <- Undefined
                62 r63 <- Undefined
                63 r64 <- Undefined
                64 r65 <- Undefined
                65 r45 <- Ancestral(AncestralName("arg0"))
                66 r14 <- Get(r45)
                67 r46 <- Ancestral(AncestralName("arg1"))
                68 r15 <- Get(r46)
                69 r47 <- Ancestral(AncestralName("arg2"))
                70 r19 <- Get(r47)
                71 r48 <- Ancestral(AncestralName("arg3"))
                72 r18 <- Get(r48)
                73 r49 <- Ancestral(AncestralName("arg4"))
                74 r20 <- Get(r49)
                75 r50 <- Ancestral(AncestralName("arg5"))
                76 r52 <- Part { src: r50, offset: 0, size: 8 }
                77 r51 <- ArithK(Add, r12, 8)
                78 r44 <- StoreMem { mem: r44, addr: r51, value: r52 }
                79 r52 <- Part { src: r50, offset: 8, size: 8 }
                80 r51 <- ArithK(Add, r12, 16)
                81 r44 <- StoreMem { mem: r44, addr: r51, value: r52 }
                82 r53 <- Ancestral(AncestralName("arg6"))
                83 r55 <- Part { src: r53, offset: 0, size: 8 }
                84 r54 <- ArithK(Add, r12, 24)
                85 r44 <- StoreMem { mem: r44, addr: r54, value: r55 }
                86 r55 <- Part { src: r53, offset: 8, size: 8 }
                87 r54 <- ArithK(Add, r12, 32)
                88 r44 <- StoreMem { mem: r44, addr: r54, value: r55 }
0x4013d2:
                89 r46 <- Const { value: 24, size: 8 }
                90 r46 <- Arith(Add, r46, r12)
                91 r45 <- LoadMem { mem: r44, addr: r46, size: 8 }
                92 r16 <- Get(r45)
0x4013d7:
                93 r46 <- Const { value: 32, size: 8 }
                94 r46 <- Arith(Add, r46, r12)
                95 r45 <- LoadMem { mem: r44, addr: r46, size: 8 }
                96 r19 <- Get(r45)
0x4013dc:
                97 r45 <- Void
                98 r45 <- Concat { lo: r45, hi: r16 }
                99 r45 <- Concat { lo: r45, hi: r19 }
               100 r45 <- Part { src: r45, offset: 0, size: 16 }
               101 r47 <- SetReturnValue(r45)
               102 r47 <- Control(Ret)


ssa pre-xform:
ssa program  103 instrs
.B0:    ;;  → Always(Return).
  (  6)  r11 <- Ancestral(AncestralName("RSP"))
         r43 <- Ancestral(AncestralName("memory"))
  (  2)  r75 <- Ancestral(AncestralName("arg5"))
         r76 <- Part { src: r75, offset: 0, size: 8 }
         r77 <- ArithK(Add, r11, 8)
         r78 <- StoreMem { mem: r43, addr: r77, value: r76 }
         r79 <- Part { src: r75, offset: 8, size: 8 }
         r80 <- ArithK(Add, r11, 16)
         r81 <- StoreMem { mem: r78, addr: r80, value: r79 }
  (  2)  r82 <- Ancestral(AncestralName("arg6"))
         r83 <- Part { src: r82, offset: 0, size: 8 }
         r84 <- ArithK(Add, r11, 24)
         r85 <- StoreMem { mem: r81, addr: r84, value: r83 }
         r86 <- Part { src: r82, offset: 8, size: 8 }
         r87 <- ArithK(Add, r11, 32)
  (  2)  r88 <- StoreMem { mem: r85, addr: r87, value: r86 }
         r89 <- Const { value: 24, size: 8 }
         r90 <- Arith(Add, r89, r11)
         r91 <- LoadMem { mem: r88, addr: r90, size: 8 }
         r93 <- Const { value: 32, size: 8 }
         r94 <- Arith(Add, r93, r11)
         r95 <- LoadMem { mem: r88, addr: r94, size: 8 }
         r98 <- Concat { lo: r97, hi: r91 }
         r99 <- Concat { lo: r98, hi: r95 }
         r100 <- Part { src: r99, offset: 0, size: 16 }
         r101 <- SetReturnValue(r100)


cfg:
  entry: BlockID(0)
  BlockID(0) -> Always(Return) [r11, r43, r75, r76, r77, r78, r79, r80, r81, r82, r83, r84, r85, r86, r87, r88, r89, r90, r91, r93, r94, r95, r97, r98, r99, r100, r101]
  domtree:
    BlockID(0)

ssa post-xform:
ssa program  113 instrs
.B0:    ;;  → Always(Return).
  (  8)  r11 <- Ancestral(AncestralName("RSP"))
         r43 <- Ancestral(AncestralName("memory"))
  (  2)  r75 <- Ancestral(AncestralName("arg5"))
         r76 <- Part { src: r75, offset: 0, size: 8 }
         r77 <- ArithK(Add, r11, 8)
         r78 <- StoreMem { mem: r43, addr: r77, value: r76 }
         r79 <- Part { src: r75, offset: 8, size: 8 }
         r80 <- ArithK(Add, r11, 16)
         r81 <- StoreMem { mem: r78, addr: r80, value: r79 }
  (  4)  r82 <- Ancestral(AncestralName("arg6"))
         r83 <- Part { src: r82, offset: 0, size: 8 }
         r84 <- ArithK(Add, r11, 24)
         r85 <- StoreMem { mem: r81, addr: r84, value: r83 }
         r86 <- Part { src: r82, offset: 8, size: 8 }
         r87 <- ArithK(Add, r11, 32)
         r88 <- StoreMem { mem: r85, addr: r87, value: r86 }
         r89 <- Const { value: 24, size: 8 }
         r90 <- ArithK(Add, r11, 24)
  (  4)  r104 <- Part { src: r82, offset: 0, size: 8 }
         r105 <- ArithK(Add, r11, 32)
  (  3)  r109 <- Part { src: r82, offset: 8, size: 8 }
         r110 <- ArithK(Add, r11, 40)
         r112 <- Get(r109)
         r107 <- Get(r104)
         r91 <- Get(r104)
         r93 <- Const { value: 32, size: 8 }
         r94 <- ArithK(Add, r11, 32)
         r95 <- Get(r109)
         r98 <- Get(r104)
  (  2)  r99 <- Concat { lo: r104, hi: r109 }
         r100 <- Get(r99)
         r101 <- SetReturnValue(r99)



T0: {
  [pre:RSP + 8]:* := pre:arg5[8 .. 0];
  [pre:RSP + 16]:* := pre:arg5[16 .. 8];
  [pre:RSP + 24]:* := pre:arg6[8 .. 0];
  [pre:RSP + 32]:* := pre:arg6[16 .. 8];
  let r104: Bytes(8) = pre:arg6[8 .. 0];
  let r109: Bytes(8) = pre:arg6[16 .. 8];
  let r99: Bytes(16) = r109 ++ r104;
  ;
  return r99
}
